### ISSUE TITLE
Developer GitHub Actons  Build & Deploy Template

### ISSUE THREAD

**Issue by mitovskaol** [BCDevOps/developer-experience#1772]
DoD:
 - A template that includes all the identified steps such as build, scan, analyze and deploy is created in [the pipeline template repo](https://github.com/bcgov/Security-pipeline-templates/tree/main/GitHubActions).

**Comment by gregnrobinson** [comment]
@mitovskaol  For building and deploying applications to OpenShift, I am proposing the following process to achieve this.

**Build**
- A Github Actions workflow for building, scanning, and pushing an image to a container repository. 

**Deploy**
*Option 1:*
- ArgoCD picks up the changes in the container repository and deploys the newly built image

*Option 2:*
- Use the Openshit Github Action to login to OpenShift and deploy the application from the Github runner. This requires repository secrets to exist in the Github repository housing the application.

ArgoCD is a good candidate for deployment because it has built-in capabilities for detecting image changes and therefore the build and deploy process. This means there is no required coordination between Github Actions and the deployment of container images into OpenShift. If Github Actions were to also handle image deployment into OpenShift, secrets need to be created in the repository housing the application. There is no wrong way of handling this and I will demonstrate both use cases.

**Comment by Nicole-Huard** [comment]
@gregnrobinson see conversation in Rocket Chat: https://chat.developer.gov.bc.ca/group/internal-platform-services/thread/CkKcQ6irmNWwyiYWu

**Comment by j-pye** [comment]
@mitovskaol started a discussion on this in RocketChat and asked me to share my thoughts from there over here.

**TL;DR** Github Actions for build + ArgoCD

Given the choice, Any pipeline tool (Github Actions, Tekton, Argo Workflows, even Jenkins) + ArgoCD is the better option. IaC / GitOps has too many benefits to ignore such as Observability, Improved stability and consistency, and an improved security model to name a few. There's so much more to it though. And when you add in tools like Vault you get a more secure model for secrets management. 

That's not to say that Github Repository Secrets aren't good but there's a lot less control and audit-ability. Plus they're supposed to be Machine Readable, Write Only secrets - What if the only person that knows the secret isn't available when you need the secret, or need to replace the secret but don't know the format? (Yes, it should have been documented but was it) - There are other concerns like public log output but github does a lot to prevent secrets from being placed in the logs unless you're actively trying to circumvent it. 

Github Actions for builds and deployments is still better than no pipeline. And better than Jenkins :upside_down: The problem is there's no guarantee that the deployment succeeded or is healthy. Custom logic can be added to the pipeline to achieve some of this but there's nothing preventing manual changes or ensuring the application stays in a specific state after the pipeline finishes. (ArgoCD ensures that, or at least tries to within reason)

**Comment by gregnrobinson** [comment]
I think I agree with everything said here. After having worked with these tools quite a bit I find Tekton requires the most development effort in that the marketplace and community is still small and the tool is modularized by design. This means if you want a functioning pipeline you need to write a task file, a pipeline file that calls the task file, and then a pipeline run file that executes the pipeline. Github actions I think is the easiest tool to use for hitting the ground running. A lot of stable actions can merely be copied into a workflow and it will have a good chance of working without problems. The widespread adoption and large community really pays to the success they have had. Tekton and Argo have the advantage of running locally in the cluster which makes it easier to manage Kubernetes resources without worrying about public endpoints and ingress requirements. Openshift ultimately takes care of this well and there are full build and deploy pipeline templates for Github actions assuming the cluster is publically accessible. 

Github has the added benefit of keeping the pipelines tightly integrated within the repository. I think this helps teams work more efficiently when using GitOps strategies. Tekton has a lot of potentials but will require more effort from teams to understand the tool and its structure. Argo I think fits smack dab in the middle as it provides a more or less complete platform that can be deployed quickly. Built-in monitoring and service breakdowns really is one thing all other tools lack out of the box. Users can choose to create their workflows using a GUI like Github or through Kubernetes manifests like Tekton. The features Argo provides for no money is quite awesome. The image-updater is another addon I like that provides automatic rollouts of new image versions without any additional configuration. Just send the new version to the registry and the image-updater pod will detect a drift and sync any services using that image to the latest version.

**Comment by Nicole-Huard** [comment]
@mitovskaol before we start to work on a recorded demo for using the GitHub Actions template, did you want your team to review what Greg has put together?

**Comment by mitovskaol** [comment]
Yes! I will ask the team and a couple of Community members to review and let you know if there is any changes required.

**Comment by Nicole-Huard** [comment]
@gregnrobinson once BC Gov team reviews, we can finalize the template, then circle back with Olena to create recordings for the community to help them get started & use the templates.

**Comment by Nicole-Huard** [comment]
@mitovskaol did any further team members want to provide feedback on the GH Actions Build & Deploy template? If not, we'll finalize and move on to the recording.

**Comment by mitovskaol** [comment]
@Nicole-Huard  Ian had a couple of questions about the template that he sent to Greg. Lets keep this ticket open for a bit longer to  give the team more time to finish the testing.

**Comment by Nicole-Huard** [comment]
No worries @mitovskaol! Once testing is completed and the tempates are updated as needed, we can close the tickets and Greg will move on to the recordings!
